File: ios-app/movie-rater/Services/MovieService.swift
import Foundation
import Combine

@MainActor
final class MovieService: ObservableObject {
    static let shared = MovieService()

    private let apiKey = "47b9a9b5aea7b85e093e203b33d41878"
    private let baseURL = "https://api.themoviedb.org/3"
    private let imageBaseURL = "https://image.tmdb.org/t/p/w500"

    @Published private(set) var movies: [Movie] = []
    @Published private(set) var isLoaded = false
    @Published private(set) var savedMovieIdsPublished: Set<Int> = []

    private init() {
        if let arr = UserDefaults.standard.array(forKey: "savedMovies") as? [Int] {
            savedMovieIdsPublished = Set(arr)
        }
        Task { await loadMovies() }
    }

    // MARK: - Favorites

    func isSaved(_ movieId: Int) -> Bool {
        savedMovieIdsPublished.contains(movieId)
    }

    func toggleSave(_ movieId: Int) {
        if savedMovieIdsPublished.contains(movieId) {
            savedMovieIdsPublished.remove(movieId)
        } else {
            savedMovieIdsPublished.insert(movieId)
        }
        UserDefaults.standard.set(Array(savedMovieIdsPublished), forKey: "savedMovies")
    }

    func savedMovies() -> [Movie] {
        movies.filter { savedMovieIdsPublished.contains($0.id) }
    }

    // MARK: - Search

    nonisolated func searchMovies(query: String) async -> [Movie] {
        guard !query.trimmingCharacters(in: .whitespaces).isEmpty else { return [] }
        let encoded = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
        let urlStr = "https://api.themoviedb.org/3/search/movie?api_key=47b9a9b5aea7b85e093e203b33d41878&language=en-US&query=\(encoded)&page=1&include_adult=false"
        guard let url = URL(string: urlStr) else { return [] }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let response = try JSONDecoder().decode(TMDbMovieResponse.self, from: data)
            var results: [Movie] = []
            for tmdb in response.results.prefix(20) {
                let movie = await Self.enrichMovie(tmdb, imageBaseURL: "https://image.tmdb.org/t/p/w500")
                results.append(movie)
            }
            return results
        } catch {
            return []
        }
    }

    // MARK: - Actor data

    nonisolated func getActorMovies(actorId: Int) async -> [Movie] {
        let urlStr = "https://api.themoviedb.org/3/person/\(actorId)/movie_credits?api_key=47b9a9b5aea7b85e093e203b33d41878"
        guard let url = URL(string: urlStr) else { return [] }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let response = try JSONDecoder().decode(TMDbActorCredits.self, from: data)
            let filtered = response.cast
                .filter { $0.poster_path != nil && ($0.release_date ?? "").isEmpty == false }
                .sorted { ($0.release_date ?? "") > ($1.release_date ?? "") }
            var results: [Movie] = []
            for tmdb in filtered {
                let movie = await Self.enrichMovie(tmdb.asMovieItem(), imageBaseURL: "https://image.tmdb.org/t/p/w500")
                results.append(movie)
            }
            return results
        } catch {
            return []
        }
    }

    nonisolated func getActor(actorId: Int) async -> Actor? {
        let urlStr = "https://api.themoviedb.org/3/person/\(actorId)?api_key=47b9a9b5aea7b85e093e203b33d41878"
        guard let url = URL(string: urlStr) else { return nil }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let person = try JSONDecoder().decode(TMDbPerson.self, from: data)
            let profileImage = person.profile_path.map { "https://image.tmdb.org/t/p/w500\($0)" } ?? ""
            return Actor(id: person.id, name: person.name, profileImage: profileImage)
        } catch {
            return nil
        }
    }

    // MARK: - Private load

    private func loadMovies() async {
        let urlStr = "\(baseURL)/movie/popular?api_key=\(apiKey)&language=en-US&page=1"
        guard let url = URL(string: urlStr) else { isLoaded = true; return }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let response = try JSONDecoder().decode(TMDbMovieResponse.self, from: data)
            var enriched: [Movie] = []
            for tmdb in response.results.prefix(20) {
                let movie = await Self.enrichMovie(tmdb, imageBaseURL: imageBaseURL)
                enriched.append(movie)
            }
            movies = enriched
        } catch {
            movies = []
        }
        isLoaded = true
    }

    // MARK: - Static helpers (nonisolated by being static)

    private static func enrichMovie(_ tmdb: TMDbMovie, imageBaseURL: String) async -> Movie {
        async let creditsTask: TMDbCredits? = fetchCredits(movieId: tmdb.id)
        async let trailerTask: String? = fetchTrailerUrl(movieId: tmdb.id)
        let (credits, trailerUrl) = await (creditsTask, trailerTask)
        return makeMovie(from: tmdb, credits: credits, trailerUrl: trailerUrl, imageBaseURL: imageBaseURL)
    }

    private static func fetchCredits(movieId: Int) async -> TMDbCredits? {
        let urlStr = "https://api.themoviedb.org/3/movie/\(movieId)/credits?api_key=47b9a9b5aea7b85e093e203b33d41878"
        guard let url = URL(string: urlStr),
              let (data, _) = try? await URLSession.shared.data(from: url) else { return nil }
        return try? JSONDecoder().decode(TMDbCredits.self, from: data)
    }

    private static func fetchTrailerUrl(movieId: Int) async -> String? {
        let urlStr = "https://api.themoviedb.org/3/movie/\(movieId)/videos?api_key=47b9a9b5aea7b85e093e203b33d41878"
        guard let url = URL(string: urlStr),
              let (data, _) = try? await URLSession.shared.data(from: url),
              let videos = try? JSONDecoder().decode(TMDbVideoResponse.self, from: data) else { return nil }
        let trailer = videos.results.first { $0.type == "Trailer" && ($0.official ?? false) }
            ?? videos.results.first { $0.type == "Trailer" }
            ?? videos.results.first { $0.type == "Teaser" }
        return trailer.map { "https://www.youtube.com/embed/\($0.key)" }
    }

    private static func makeMovie(from tmdb: TMDbMovie, credits: TMDbCredits?, trailerUrl: String?, imageBaseURL: String) -> Movie {
        let year: Int
        if let dateStr = tmdb.release_date, !dateStr.isEmpty {
            let formatter = DateFormatter()
            formatter.dateFormat = "yyyy-MM-dd"
            year = formatter.date(from: dateStr).flatMap { Calendar.current.dateComponents([.year], from: $0).year } ?? Calendar.current.component(.year, from: Date())
        } else {
            year = Calendar.current.component(.year, from: Date())
        }
        let cast = credits?.cast.prefix(10) ?? []
        return Movie(
            id: tmdb.id,
            title: tmdb.title,
            image: tmdb.poster_path.map { "\(imageBaseURL)\($0)" } ?? "",
            description: tmdb.overview ?? "No description available.",
            actors: cast.map { $0.name },
            actorImages: cast.map { $0.profile_path.map { "\(imageBaseURL)\($0)" } ?? "" },
            actorIds: cast.map { $0.id },
            year: year,
            rating: tmdb.vote_average ?? 0,
            link: "https://www.themoviedb.org/movie/\(tmdb.id)",
            trailerUrl: trailerUrl
        )
    }
}

// MARK: - TMDb DTOs

private struct TMDbMovieResponse: Decodable {
    let results: [TMDbMovie]
}

private struct TMDbMovie: Decodable {
    let id: Int
    let title: String
    let poster_path: String?
    let overview: String?
    let release_date: String?
    let vote_average: Double?
}

private struct TMDbCredits: Decodable {
    let cast: [TMDbCastMember]
}

private struct TMDbCastMember: Decodable {
    let id: Int
    let name: String
    let profile_path: String?
}

private struct TMDbVideoResponse: Decodable {
    let results: [TMDbVideo]
}

private struct TMDbVideo: Decodable {
    let key: String
    let type: String
    let official: Bool?
}

private struct TMDbActorCredits: Decodable {
    let cast: [TMDbActorMovie]
}

private struct TMDbActorMovie: Decodable {
    let id: Int
    let title: String
    let poster_path: String?
    let overview: String?
    let release_date: String?
    let vote_average: Double?

    func asMovieItem() -> TMDbMovie {
        TMDbMovie(id: id, title: title, poster_path: poster_path, overview: overview, release_date: release_date, vote_average: vote_average)
    }
}

private struct TMDbPerson: Decodable {
    let id: Int
    let name: String
    let profile_path: String?
}

File: android-app/app/src/main/java/com/example/movierater/data/MovieService.kt
package com.example.movierater.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

private const val TMDB_API_KEY = "47b9a9b5aea7b85e093e203b33d41878"
private const val TMDB_BASE_URL = "https://api.themoviedb.org/3/"
private const val TMDB_IMAGE_BASE_URL = "https://image.tmdb.org/t/p/w500"

private val json = Json { ignoreUnknownKeys = true }

private val httpClient = OkHttpClient.Builder()
    .addInterceptor(HttpLoggingInterceptor().apply {
        level = HttpLoggingInterceptor.Level.BASIC
    })
    .build()

private val retrofit = Retrofit.Builder()
    .baseUrl(TMDB_BASE_URL)
    .client(httpClient)
    .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
    .build()

private val tmdbApi = retrofit.create(TMDbApi::class.java)

interface TMDbApi {
    @GET("movie/popular")
    suspend fun getPopularMovies(
        @Query("api_key") apiKey: String = TMDB_API_KEY,
        @Query("language") language: String = "en-US",
        @Query("page") page: Int = 1,
    ): MovieResponse

    @GET("search/movie")
    suspend fun searchMovies(
        @Query("api_key") apiKey: String = TMDB_API_KEY,
        @Query("language") language: String = "en-US",
        @Query("query") query: String,
        @Query("page") page: Int = 1,
        @Query("include_adult") includeAdult: Boolean = false,
    ): MovieResponse

    @GET("movie/{movieId}/credits")
    suspend fun getCredits(
        @Path("movieId") movieId: Int,
        @Query("api_key") apiKey: String = TMDB_API_KEY,
    ): Credits

    @GET("movie/{movieId}/videos")
    suspend fun getVideos(
        @Path("movieId") movieId: Int,
        @Query("api_key") apiKey: String = TMDB_API_KEY,
    ): VideoResponse

    @GET("person/{personId}")
    suspend fun getPerson(
        @Path("personId") personId: Int,
        @Query("api_key") apiKey: String = TMDB_API_KEY,
    ): Person

    @GET("person/{personId}/movie_credits")
    suspend fun getActorMovies(
        @Path("personId") personId: Int,
        @Query("api_key") apiKey: String = TMDB_API_KEY,
    ): ActorCredits
}

@Serializable
data class MovieResponse(
    val results: List<TMDbMovie>,
)

@Serializable
data class TMDbMovie(
    val id: Int,
    val title: String,
    @SerialName("poster_path")
    val posterPath: String?,
    val overview: String?,
    @SerialName("release_date")
    val releaseDate: String?,
    @SerialName("vote_average")
    val voteAverage: Double?,
)

@Serializable
data class Credits(
    val cast: List<CastMember>,
)

@Serializable
data class CastMember(
    val id: Int,
    val name: String,
    @SerialName("profile_path")
    val profilePath: String?,
)

@Serializable
data class VideoResponse(
    val results: List<Video>,
)

@Serializable
data class Video(
    val key: String,
    val type: String,
    val official: Boolean? = false,
)

@Serializable
data class Person(
    val id: Int,
    val name: String,
    @SerialName("profile_path")
    val profilePath: String?,
)

@Serializable
data class ActorCredits(
    val cast: List<ActorMovie>,
)

@Serializable
data class ActorMovie(
    val id: Int,
    val title: String,
    @SerialName("poster_path")
    val posterPath: String?,
    val overview: String?,
    @SerialName("release_date")
    val releaseDate: String?,
    @SerialName("vote_average")
    val voteAverage: Double?,
)

object MovieService {
    suspend fun getPopularMovies(): List<Movie> {
        return try {
            val response = tmdbApi.getPopularMovies()
            response.results.take(20).map { tmdbMovie ->
                enrichMovie(tmdbMovie)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun searchMovies(query: String): List<Movie> {
        if (query.isBlank()) return emptyList()
        return try {
            val response = tmdbApi.searchMovies(query = query)
            response.results.take(20).map { tmdbMovie ->
                enrichMovie(tmdbMovie)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getActorMovies(actorId: Int): List<Movie> {
        return try {
            val response = tmdbApi.getActorMovies(actorId)
            response.cast
                .filter { it.posterPath != null && it.releaseDate?.isNotEmpty() == true }
                .sortedByDescending { it.releaseDate }
                .take(20)
                .map { actorMovie ->
                    enrichMovie(
                        TMDbMovie(
                            id = actorMovie.id,
                            title = actorMovie.title,
                            posterPath = actorMovie.posterPath,
                            overview = actorMovie.overview,
                            releaseDate = actorMovie.releaseDate,
                            voteAverage = actorMovie.voteAverage,
                        )
                    )
                }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getActor(actorId: Int): Actor? {
        return try {
            val person = tmdbApi.getPerson(actorId)
            Actor(
                id = person.id,
                name = person.name,
                profileImage = person.profilePath?.let { "$TMDB_IMAGE_BASE_URL$it" },
            )
        } catch (e: Exception) {
            null
        }
    }

    private suspend fun enrichMovie(tmdbMovie: TMDbMovie): Movie {
        val (credits, videos) = coroutineScope {
            val creditsDeferred = async {
                try {
                    tmdbApi.getCredits(tmdbMovie.id)
                } catch (e: Exception) {
                    null
                }
            }
            val videosDeferred = async {
                try {
                    tmdbApi.getVideos(tmdbMovie.id)
                } catch (e: Exception) {
                    null
                }
            }
            Pair(creditsDeferred.await(), videosDeferred.await())
        }

        val cast = credits?.cast?.take(10) ?: emptyList()
        val trailerUrl = videos?.results?.let { videoList ->
            (videoList.firstOrNull { it.type == "Trailer" && (it.official ?: false) }
                ?: videoList.firstOrNull { it.type == "Trailer" }
                ?: videoList.firstOrNull { it.type == "Teaser" })?.let {
                "https://www.youtube.com/embed/${it.key}"
            }
        }

        val year = extractYear(tmdbMovie.releaseDate)

        return Movie(
            id = tmdbMovie.id,
            title = tmdbMovie.title,
            image = tmdbMovie.posterPath?.let { "$TMDB_IMAGE_BASE_URL$it" } ?: "",
            description = tmdbMovie.overview ?: "No description available.",
            actors = cast.map { it.name },
            actorImages = cast.map { it.profilePath?.let { path -> "$TMDB_IMAGE_BASE_URL$path" } ?: "" },
            actorIds = cast.map { it.id },
            year = year,
            rating = tmdbMovie.voteAverage ?: 0.0,
            link = "https://www.themoviedb.org/movie/${tmdbMovie.id}",
            trailerUrl = trailerUrl,
        )
    }

    private fun extractYear(dateStr: String?): Int {
        if (dateStr?.isNotEmpty() == true) {
            return try {
                val formatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val date = formatter.parse(dateStr)
                if (date != null) {
                    val calendar = Calendar.getInstance().apply { time = date }
                    calendar.get(Calendar.YEAR)
                } else {
                    Calendar.getInstance().get(Calendar.YEAR)
                }
            } catch (e: Exception) {
                Calendar.getInstance().get(Calendar.YEAR)
            }
        }
        return Calendar.getInstance().get(Calendar.YEAR)
    }
}
